NDTensors/ext/NDTensorCUDA/set_types.jl

```diff
@@ -0,0 +1,34 @@
+buffertype(datatype::Type{<:CuArray{<:Any,<:Any,B}}) where {B} = B
+function buffertype(datatype::Type{<:CuArray})
+  println(
+    "CuArray definitions require a CUDA.Mem buffer try $(datatype{default_buffertype()})"
+  )
+  throw(TypeError)
```

Comment:
I think `$(datatype{default_buffertype()})` won't always expand to what you hope it will here, since if someone just passes `CuArray` it would make it look like the buffer type should be passed as the first type parameter.

Also, I don't think you're using `TypeError` correctly, it would be used like this:
```julia
julia> throw(TypeError(:typeassert, Float64, 1))
ERROR: TypeError: in typeassert, expected Float64, got a value of type Int64
Stacktrace:
 [1] top-level scope
   @ REPL[46]:1
```
which gets output by type assertions such as:
```julia
julia> 1::Float64
ERROR: TypeError: in typeassert, expected Float64, got a value of type Int64
Stacktrace:
 [1] top-level scope
   @ REPL[43]:1
```

I would go with something simpler, like:
```julia
function buffertype(datatype::Type{<:CuArray})
  return error("No buffer type specified in the type $(datatype)")
end
```